tdms/src/interpolate_Efield.cpp

```diff
@@ -1,66 +1,113 @@
 
+using namespace std;
 
 void interpolateTimeDomainEx(double ***Exy, double ***Exz, int i, int j, int k, int nI, double *Ex) {
-
     // determine the interpolation scheme to use
     const interpScheme &scheme = best_interp_scheme(nI, i);
     // prepare input data - if using a cubic scheme we have reserved more memory than necessary but nevermind
     double interp_data[8];
 
     // now fill the interpolation data
     // i - (scheme.number_of_datapoints_to_left) is the index of the Yee cell that plays the role of v0 in the interpolation
-    for(int ind=scheme.first_nonzero_coeff; ind<=scheme.last_nonzero_coeff; ind++) {
+    for (int ind = scheme.first_nonzero_coeff; ind <= scheme.last_nonzero_coeff; ind++) {
         interp_data[ind] = Exy[k][j][i - scheme.number_of_datapoints_to_left + ind] + Exz[k][j][i - scheme.number_of_datapoints_to_left + ind];
     }
 
     // now run the interpolation scheme and place the result into the output
     *Ex = scheme.interpolate(interp_data);
 }
+void interpolateFreqDomainEx(complex<double> ***Ex, int i, int j, int k, int nI, complex<double> *Ex_interp) {
+    // determine the interpolation scheme to use
+    const interpScheme &scheme = best_interp_scheme(nI, i);
+    // prepare input data - if using a cubic scheme we have reserved more memory than necessary but nevermind
+    std::complex<double> interp_data[8];
 
-void interpolateTimeDomainEy(double ***Eyx, double ***Eyz, int i, int j, int k, int nJ, double *Ey)
-{
+    // now fill the interpolation data
+    // i - (scheme.number_of_datapoints_to_left) is the index of the Yee cell that plays the role of v0 in the interpolation
+    for (int ind = scheme.first_nonzero_coeff; ind <= scheme.last_nonzero_coeff; ind++) {
+        interp_data[ind] = Ex[k][j][i - scheme.number_of_datapoints_to_left + ind];
+    }
+    // now run the interpolation scheme and place the result into the output
+    *Ex_interp = scheme.interpolate(interp_data);
+}
+
+void interpolateTimeDomainEy(double ***Eyx, double ***Eyz, int i, int j, int k, int nJ, double *Ey) {
     // determine the interpolation scheme to use
     const interpScheme &scheme = best_interp_scheme(nJ, j);
     // prepare input data - if using a cubic scheme we have reserved more memory than necessary but nevermind
     double interp_data[8];
 
     // now fill the interpolation data
     // j - scheme.number_of_datapoints_to_left is the index of the Yee cell that plays the role of v0 in the interpolation
-    for (int ind = scheme.first_nonzero_coeff; ind <= scheme.last_nonzero_coeff; ind++)
-    {
-        interp_data[ind] = Eyx[k][j - scheme.number_of_datapoints_to_left + ind][i] + Eyz[k][j - scheme.number_of_datapoints_to_left + ind][j];
+    for (int ind = scheme.first_nonzero_coeff; ind <= scheme.last_nonzero_coeff; ind++) {
+        interp_data[ind] = Eyx[k][j - scheme.number_of_datapoints_to_left + ind][i] + Eyz[k][j - scheme.number_of_datapoints_to_left + ind][i];
```

Comment:
This is easiest to explain in pictures - grab me in BH/HH next week if what's below isn't clear. Some setup below:

General information about the algorithm is in the `PSTD_interpolation_scheme` document. The caviat is that the field values associated to the Yee cells lie on the faces of the cells, rather than at their centres (hence the need for interpolation). So the field value stored in `Eyx[k][j][i]` is the field value _associated_ to the cell `(i,j,k)`, _but_ isn't the value of the field at the Yee cell centre. In fact, if `(x_i, y_j, z_k)` are the coordinates of the centre of cell `(i,j,k)`, then `Eyx[k][j][i]` is the split-field value at position `(x_i, y_j + delta_j, z_k)` where `delta_j` is half the length in the $y$-direction of the Yee cell. There are similar rules in Peter's document for the other field components.

As such, we interpolate "between the Yee cell _values_" (the `Eyx[k][j][i]` etc) to obtain the field at the Yee cell centres. Since the `Eyx[k][j][i]` are all evenly spaced, we have a situation like below:
```
X     X     X     X     X     X     X     X
                     o
```
Here, the `X`s are the positions at which we have the `Eyx[k][j][i]`, whilst `o` is the centre of one particular Yee cell that we want to interpolate into. Since the `Eyx[k][j][i]` are evenly-spaced, and `o` is at the midpoint of two `X`s (being the centre of a particular Yee cell), we can interpolate using the Yee cell values "at the edges" (the `Eyx[k][j][i]`) to obtain the fields at the centre of the Yee cells.

Now the answer to your question :sweat_smile: 
We're interpolating to the centre of cell `(i,j,k)` in the `y`-direction, so we don't change the `k` and `i` indices in `Eyx` when we extract the datapoints we need. Let's suppose we have a situation like the above, so we then need to look at 4 datapoints either side of the `o`. But remember that the datapoint immediately to the _right_ of `o` is the value `Eyx[k][j][i]`. So, the 8 datapoints we need must come from the Yee cells whose y-indices are:
```
j-4, j-3, j-2, j-1,   j , j+1, j+2, j+3
 X    X    X    X     X    X    X    X
                   o
```
Hence, this weird `for` loop. `interp_data` is an array of length 8 that'll store the datapoints that we use to interpolate with. `ind` is an index running from 0 to 7 that we can use to extract the 8 consecutive values from `Eyx` we need. `scheme.number_of_datapoints_to_left` in this instance would be 4 - we need 4 datapoints (`j-4` to `j-1`) before `Eyx[k][j][i]` to perform this scheme.

The `InterpScheme` class sets the private value of `number_of_datapoints_to_left` when an instance is created, because in general we might be in a situation where we don't have 4 cells either side of `o` to use in our interpolation. For example,
```
X     X     X     X     X     X     X     X
         o             
```
in the above, `number_of_datapoints_to_left` will be 2, and we will take datapoints from cells `j-2` through to `j+5`. Conversely, a situation like this
```
X     X     X     X     X     X     X     X
                                       o
```
will have `number_of_datapoints_to_left` being 7.

Finally, you also highlighted that `ind` doesn't go from 0 to 7, but rather from `scheme.first_nonzero_coeff` to `scheme.last_nonzero_coeff`. Again, these are just fail-safes for when we are in these non-symmetric situations. When we are near the edges, some interpolation schemes require fewer than 8 datapoints (since the some of the scheme's coefficients are 0), so this prevents seg-faults that might happen if we _always_ tried to get 4 datapoints either side of `o`. It also means that when we fall back on cubic interpolation, we don't always try and read double as many values as we need!